fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Reference.java

```diff
@@ -111,185 +179,237 @@ public RecordQueryPlan getAsPlan() {
+        return exploratoryMembers.isEmpty() ? finalMembers.getOnlyElement() : exploratoryMembers.getOnlyElement();
     }
 
     /**
      * Legacy method that calls {@link #pruneWith(RelationalExpression)} while being synchronized on {@code this}.
-     * @param newValue new expression to replace members of this reference.
+     * @param plan new expression to replace members of this reference.
      */
-    public synchronized void replace(@Nonnull RelationalExpression newValue) {
-        pruneWith(newValue);
+    public synchronized void replace(@Nonnull final RecordQueryPlan plan) {
+        pruneWithUnchecked(plan);
     }
 
     /**
      * Method that replaces the current members of this reference with a new value. This is called by the planner
      * to prune the variations of a reference down to exactly one new member.
-     * @param newValue new value to replace existing members
+     * @param expression expression to replace existing members with
      */
-    public void pruneWith(@Nonnull RelationalExpression newValue) {
-        final Map<ExpressionProperty<?>, ?> propertiesForPlan;
-        if (newValue instanceof RecordQueryPlan) {
-            propertiesForPlan = propertiesMap.getCurrentProperties(newValue);
-        } else {
-            propertiesForPlan = null;
-        }
-        clear();
-        insertUnchecked(newValue, propertiesForPlan);
+    public void pruneWith(@Nonnull final RelationalExpression expression) {
+        Verify.verify(isFinal(expression));
+        pruneWithUnchecked(expression);
+    }
+
+    private void pruneWithUnchecked(@Nonnull final RelationalExpression expression) {
+        final var properties = propertiesMap.getCurrentProperties(expression);
+        clearFinalExpressions();
+        insertUnchecked(expression, true, properties);
     }
 
     /**
-     * Inserts a new expression into this reference. This particular overload utilized the precomputed propertied of
-     * a {@link RecordQueryPlan} that already is thought to reside in another {@link Reference}.
-     * @param newValue new expression to be inserted
-     * @param otherRef a reference that already contains {@code newValue}
+     * Inserts a new expression into this reference. This particular overload utilized the precomputed properties of
+     * a final {@link RelationalExpression} that already is thought to reside in another {@link Reference}.
+     * @param expression new expression to be inserted
+     * @param otherRef a reference that already contains {@code expression}
      * @return {@code true} if and only if the new expression was successfully inserted into this reference, {@code false}
      *         otherwise.
      */
-    public boolean insertFrom(@Nonnull final RelationalExpression newValue, @Nonnull final Reference otherRef) {
-        if (newValue instanceof RecordQueryPlan) {
+    public boolean insertFrom(@Nonnull final RelationalExpression expression, @Nonnull final Reference otherRef) {
```

Comment:
This method is currently unused. Do you have plans for it in the future?

Reply:
Removed all currently unused methods in `Reference`. We can add them back later if needed.